### DISCUSSION TITLE
[Kamal 2] How to use env vars or secrets for hosts/server ip??

### DISCUSSION THREAD

**Discussion by jgmontoya** [basecamp/kamal#1137] [Q&A] [answered]
In kamal 1, having multiple destinations (default, production), I'd have something like this:

```yml
# deploy.yml
servers:
  web:
    hosts:
      - <%= ENV['SERVER_IP_ADDRESS'] %>
```

Then in the `.env` and `.env.production` I'd have different values for `SERVER_IP_ADDRESS`

I'm having a hard time trying to replicate this on kamal 2.

If I include the `<% require "dotenv"; Dotenv.load(".env") %>` I can make it work for the default destination, but even if I add `<% require "dotenv"; Dotenv.load(".env.production") %>` to the `.env.production` I still end up with the value from `.env`

I've also tried getting it from the secrets: adding `SERVER_IP_ADDRESS` to both `.kamal/secrets` and `.kamal/secrets.production`, adding `SERVER_IP_ADDRESS` under the `env.secret` key on the `deploy.yml` and trying to set it like:

```yml
# deploy.yml
servers:
  web:
    hosts:
      - SERVER_IP_ADDRESS
```

Yet that results with the literal string `'SERVER_IP_ADDRESS'` instead of the actual secret being used.

What's the right way of doing this in kamal 2?

**Comment by ehutzelman** [comment]
A couple things you could try to get around this. 

If using multiple destinations, you can use dedicated deploy config files for each: `deploy.yml` and `deploy.production.yml` and just set the IPs directly in those files.

Like you've discovered, kamal doesn't pull from your `.env*` files anymore without some extra config. If you want to keep the IPs in env and reference from deploy.yml, I think it's more straightforward to run dotenv directly from the command line like:

`bundle exec dotenv kamal secrets print`
`bundle exec dotenv -f '.env.production' kamal secrets print`

This way you can control what `.env*` file you are using on the fly and not have to hardcode in your `deploy.yml`.

**Comment by jgmontoya** [comment]
> If using multiple destinations, you can use dedicated deploy config files for each: deploy.yml and deploy.production.yml and just set the IPs directly in those files.

I'm trying to avoid hard-coding the IPs on the deploy files

> If you want to keep the IPs in env and reference from deploy.yml, I think it's more straightforward to run dotenv directly from the command line like:
> 
> bundle exec dotenv kamal secrets print
> bundle exec dotenv -f '.env.production' kamal secrets print
> 
> This way you can control what .env* file you are using on the fly and not have to hardcode in your deploy.yml.

I'm not sure I follow, how should I then set the deploy.yml file for this to work? How do I 'tell' kamal that it needs to run dotenv for its other commands (deploy, redeploy, etc..) to work?

**Reply by ehutzelman** [reply]
It should work the way you have it originally:
```
# deploy.yml
servers:
  web:
    hosts:
      - <%= ENV['SERVER_IP_ADDRESS'] %>
```
Then deploy with
```
bundle exec dotenv kamal deploy
bundle exec dotenv -f '.env.production' kamal deploy -d production
```
Assuming SERVER_IP_ADDRESS is defined in you .env files. You would have to preface all your kamal commands with the dotenv command if they need the ENV vars. If you don't want to do this, you could go back to putting the require/dotenv load back in the deploy files themselves, but I think you would need to differentiate the .env file you wanted:
```
#deploy.yml
<% require "dotenv"; Dotenv.load(".env") %>
...
```
```
#deploy.production.yml
<% require "dotenv"; Dotenv.load(".env.production") %>
...
```
Then you could just run `kamal deploy` and `kamal deploy -d production`

**Comment by jgmontoya** [comment, marked as answer]
It was a matter of using the `overwrite: true` option on the production config:

```yml
<% require "dotenv"; Dotenv.load(".env.production", overwrite: true) %>
...
```

This will make .env.production keys have precedence over .env keys
